src/Components/Components/src/Rendering/ComponentState.cs

```diff
@@ -222,5 +228,38 @@ private void DisposeBuffers()
             ((IDisposable)CurrentRenderTree).Dispose();
             _latestDirectParametersSnapshot?.Dispose();
         }
+
+        internal bool RequiresAsyncDisposal() => Component is IAsyncDisposable;
+
+        internal Task DisposeInBatchAsync(RenderBatchBuilder batchBuilder)
+        {
+            _componentWasDisposed = true;
+
+            CleanupComponentStateResources(batchBuilder);
+
+            if (Component is IAsyncDisposable asyncDisposable)
```

Comment:
Why check this? We already know it's true because the caller wouldn't be on this code path otherwise.

Can we replace this with a `Debug.Assert(Component is IAsyncDisposable)`? Or even just do this so the cast would throw:

```cs
// Caller is responsible for the invariant that you can only call this if it's an async disposable
var asyncDisposable = (IAsyncDisposable)Component;
```

I still wonder about whether we could unify the sync/async disposal paths a bit more. For example, `ComponentState` *only* exposed a `DisposeInBatchAsync`, which just returned `Task.CompletedTask` for non-async-disposables. Might this even simplify the calling side for `Renderer`, because then it doesn't have to care about which disposal type we're doing - it just has to handle both synchronous and asynchronous completion of the task (i.e., don't await or do anything else asynchronously if it already completed synchronously), which it has to do anyway if it's handling the async case at all.

Reply:
I've simpliefied the thing. I would be happy if we agree to dispose the buffers, etc from the component state before calling dispose in any of their variants. I believe there is no functional difference on the sync case, but that's one thing I wanted to get your opinion on.

The other reason I didn't fuse them was to avoid an allocation in the syncrhonous case if there was an exception, since it would have to be wrapped in a task for consistency.

I don't think its super important that they are separate methods, specially given that now they share the core.